cleanup JD Protocol specs (#104)
stratum-mining/sv2-spec: 06-Job-Declaration-Protocol.md

# 6. Job Declaration Protocol

The Job Declaration Protocol is used to coordinate the creation of custom work, avoiding scenarios where Pools are unilaterally imposing work on miners.

Pools that opt into this protocol are only responsible for accounting shares and distributing rewards.

This is a key feature of Stratum V2 that improves Bitcoin decentralization.

## 6.1 Job Declarator Server

The Job Declarator Server (JDS) is deployed on Pool side, although it could be theoretically outsourced to a third party that is trusted by the Pool.

> In order to fully implement the Server side of the Job Declaration Protocol, the JDC also needs to exchange RPCs with a Bitcoin Node. 

Reviewer: ```suggestion
In order to fully implement the Server side of the Job Declaration Protocol, the JDS also needs to exchange RPCs (or similar) with a Bitcoin Node. 
```

It is responsible for:
- Allocating tokens that JDC will use to declare Custom Jobs.
- Acknowledging declaration of Custom Jobs associated with specific allocated tokens.

> - Maintaining an internal mempool (via RPCs to a Bitcoin Node).

Reviewer: ```suggestion
- Maintaining an internal mempool (via RPCs (or similar) to a Bitcoin Node).
```

- Requesting identification for transactions on some declared Custom Job.
- Requesting missing transactions on some declared Custom Job.
- Publishing valid block submissions received from JDC.

## 6.2 Job Declarator Client

The Job Declarator Client (JDC) is deployed on the miner side.

> In order to fully implement the Client side of the Job Declaration Protocol, the JDS also needs to operate under the Template Distribution and Mining Protocols.

Reviewer: ```suggestion
In order to fully implement the Client side of the Job Declaration Protocol, the JDC also needs to operate under the Template Distribution and Mining Protocols.
```

It is responsible for:
- Receiving Templates from the Template Provider (via Template Distribution Protocol).
- Declaring Custom Jobs to JDS (via Job Declaration Protocol).
- Notifying declared Custom Jobs to Pool (via Mining Protocol).

> - Receiving Shares from downstream Mining Devices working on Custom Jobs (via Mining Protocol)..

Reviewer: ```suggestion
- Receiving Shares from downstream Mining Devices working on Custom Jobs (via Mining Protocol).
```

- Submitting Shares for Custom Jobs to Pool.

> - Publishing valid blocks found by downstream Mining Devices (both to TP and JDS).

Reviewer: ```suggestion
- Publishing valid blocks found by downstream Mining Devices (both to Template Provider (TP) and JDS).
```

No flags are yet defined for use in `SetupConnection.Success`.

> ### 6.3.2 `AllocateMiningJobToken` (Client -> Server)

Author: ```suggestion
### 6.3.2 `AllocateMiningJobToken` (JDC -> JDS)
```

| async_mining_allowed                | BOOL      | If true, the mining_job_token can be used immediately on a mining connection in the SetCustomMiningJob message, even before DeclareMiningJob and DeclareMiningJob.Success messages have been sent and received. If false, Job Declarator MUST use this token for DeclareMiningJob only. <br>This MUST be true when SetupConnection.flags had REQUIRES_ASYNC_JOB_MINING set. |
| coinbase_tx_outputs         | B0_64K         | Bitcoin transaction outputs added by the pool                                                                            |

### 6.3.4 `DeclareMiningJob` (Client -> Server)

A request sent by JDC that proposes a selected set of transactions to JDS.

[`SipHash`](https://www.aumasson.jp/siphash/siphash.pdf) is used for short txids as a strategy to reduce bandwidth consumption, since including full txids (transaction data hash) would make the message body excessively big.

> More specifically, the `SipHash 2-4` variant is used. Aside from the preimage (which is the full txid), keys `k0` and `k1` are used to avoid collisions, which is the main purpose of `tx_short_hash_nonce`.

Author: related to short ids, which will be dropped from the protocol

I wouldn't spend too much time reading about it tbh

| async_mining_allowed                | BOOL      | If true, the mining_job_token can be used immediately on a mining connection in the SetCustomMiningJob message, even before DeclareMiningJob and DeclareMiningJob.Success messages have been sent and received. If false, Job Declarator MUST use this token for DeclareMiningJob only. <br>This MUST be true when SetupConnection.flags had REQUIRES_ASYNC_JOB_MINING set. |
| coinbase_tx_outputs         | B0_64K         | Bitcoin transaction outputs added by the pool                                                                            |

### 6.4.4 `DeclareMiningJob` (Client -> Server)

A request sent by JDC that proposes a selected set of transactions to JDS.

> [`SipHash`](https://www.aumasson.jp/siphash/siphash.pdf) is used for short txids as a strategy to reduce bandwidth consumption, since including full txids (transaction data hash) would make the message body excessively big.

Author: remove this

| request_id           | U32       | Identifier of the original request                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  |
| new_mining_job_token | B0_255    | Unique identifier provided by the pool of the job that the Job Declarator has declared with the pool. It MAY be the same token as DeclareMiningJob::mining_job_token if the pool allows to start mining on not yet declared job. If the token is different from the one in the corresponding DeclareMiningJob message (irrespective of if the client is already mining using the original token), the client MUST send a SetCustomMiningJob message on each Mining Protocol client which wishes to mine using the declared job. |

### 6.3.6 `DeclareMiningJob.Error` (Server->Client)

> 

Author: well, I guess it is worth pointing that out, but we should avoid generalizations

what if there was something wrong with `DeclareMiningJob`?

it should be stressed that this should trigger fallback only if this is a response to a `DeclareMiningJob` that JDC deems to be valid

> 

Author: I see your point.

I pushed some updates addressing all the comments, including this one.
